common/rfb/CMsgWriter.cxx

```diff
@@ -176,15 +176,37 @@ void CMsgWriter::writeKeyEvent(uint32_t keysym, uint32_t keycode, bool down)
 void CMsgWriter::writePointerEvent(const Point& pos, uint16_t buttonMask)
 {
   Point p(pos);
+  bool extendedMouseButtons;
+
   if (p.x < 0) p.x = 0;
   if (p.y < 0) p.y = 0;
   if (p.x >= server->width()) p.x = server->width() - 1;
   if (p.y >= server->height()) p.y = server->height() - 1;
 
+  // Only send extended pointerEvent message when needed
+  extendedMouseButtons = buttonMask & 0x180;
+
   startMsg(msgTypePointerEvent);
-  os->writeU8(buttonMask);
-  os->writeU16(p.x);
-  os->writeU16(p.y);
+  if (server->supportsExtendedMouseButtons && extendedMouseButtons) {
+    int higherBits;
+    int lowerBits;
+
+    higherBits = (buttonMask  >> 7) & 0xff;
+    lowerBits = buttonMask & 0x7f;
+    lowerBits |= 0x80; // Set marker bit to 1
+
+    higherBits &= 0x03; // Clear reserved bits
```

Comment:
I think asserts are better at this layer.

But mapping is handled by the caller, so should we have this here at all?

Reply:
Good point, I've changed it to an assert instead. Not sure if it fits better here or in the Viewport code (or somewhere in-between)